MaxText/inference/page_manager.py

```diff
@@ -1,232 +1,582 @@
-  max_pages_per_slot: int
-
-  def __init__(self, num_pages: int, tokens_per_page: int, max_target_length: int, max_prefill_length: int, batch_size: int):
-    self.num_pages = num_pages
-    self.tokens_per_page = tokens_per_page
-    self.max_target_length = max_target_length
-    self.max_pages_per_slot = max_target_length // tokens_per_page
-    self.slots = int(batch_size * jax.device_count())
-    self.max_prefill_predict_length = max_prefill_length
-
-  def release_slot_pages(
-      self,
-      slot: int,
-      page_state: PageState,
-  ) -> PageState:
-    page_status = page_state.page_status
-    page_map = page_state.page_map
-    sequence_lengths = page_state.sequence_lengths
-    current_page = page_state.current_page
-    current_page_position = page_state.current_page_position
-    num_pages_used = page_state.num_pages_used
-
-    def _release_page(i, state):
-      page_map, page_status = state
-      page_idx = page_map[slot][i]
-      page_status = page_status.at[page_idx].set(0)
-      page_map = page_map.at[slot, i].set(0)
-      return page_map, page_status
-
-    page_map, page_status = jax.lax.fori_loop(0, num_pages_used[slot], _release_page, (page_map, page_status))
-
-    sequence_lengths = sequence_lengths.at[slot].set(0)
-    num_pages_used = num_pages_used.at[slot].set(0)
-    current_page = current_page.at[slot].set(0)
-    current_page_position = current_page_position.at[slot].set(0)
-
-    return PageState(
-        page_status=page_status,
-        page_map=page_map,
-        sequence_lengths=sequence_lengths,
-        num_pages_used=num_pages_used,
-        current_page=current_page,
-        current_page_position=current_page_position,
-    )
 
-  def reserve_prefix_slot_pages(
-      self,
-      slot: int,
-      true_length: int,
-      page_state: PageState,
-  ) -> PageState:
-    page_state = self.release_slot_pages(slot, page_state)
-    page_status = page_state.page_status
-    page_map = page_state.page_map
-    sequence_lengths = page_state.sequence_lengths
-    num_pages_used = page_state.num_pages_used
-    current_page = page_state.current_page
-    current_page_position = page_state.current_page_position
-
-    prefill_slot_num_pages = jnp.ceil(true_length / self.tokens_per_page).astype(jnp.int32)
-    prefill_slot_page_slice_idx = jnp.where(true_length == 0, 0, (true_length - 1) % self.tokens_per_page)
-
-    def _reserve_page(i, state):
-      slot, page_map, page_status, current_page = state
-      page_idx = jnp.where((page_status[1:] == 0), size=1)[0][0] + 1
-      page_status = page_status.at[page_idx].set(1)
-      page_map = page_map.at[slot, i].set(page_idx)
-      current_page = current_page.at[slot].set(page_idx)
-      return slot, page_map, page_status, current_page
-
-    _, page_map, page_status, current_page = jax.lax.fori_loop(
-        0, prefill_slot_num_pages, _reserve_page, (slot, page_map, page_status, current_page)
+def _update_prefill_pages_for_group(
```

Comment:
nit: consider naming it release_and_reserve_for_group to be explicit.

Reply:
I will rename it. 

We would still need to call a release somewhere before doing the reserve in order to ensure we are keeping the PageState accurate. Can you explain a bit more about what you have in mind for this?

Reply:
From API design aspect, it is more natural to me to have two apis. One for reserve and the other for release. As we have a release_page API above, I expect this API to be only perform reservation. 

Alternatively we can have 3 APIs, one for reserve, one for release and another one for release_and_reserve. 

IMHO, it looks to me okay to require callsites to invoke release first and then reserve, thus no need for an API/method to do release_and_reserve. 

This is nitpicking. Feel free to leave as is if you feel strongly about doing this week

Reply:
I will change the name, but leave the functionality as-is for now in order to finish the PR. I will add a todo and fix this in the next round of cleanup.

Reply:
I went ahead and changed this to have individual release and reserve page functions and then left the `release_and_reserve` to call both.